src/mailmap.c

```diff
@@ -14,195 +14,312 @@
+	char_range *replace_name,
+	char_range *replace_email)
 {
-	const char* email_start;
-	const char* email_end;
-	const char* name_start;
-	const char* name_end;
+	char_range line, comment, name_a, email_a, name_b, email_b;
+	bool two_emails = false;
+
+	*found = false;
+	*real_name = NULL_RANGE;
+	*real_email = NULL_RANGE;
+	*replace_name = NULL_RANGE;
+	*replace_email = NULL_RANGE;
+
+	while (file->len) {
+		/* Get the line, and remove any comments */
+		range_split(*file, '\n', &line, file);
+		range_split(line, '#', &line, &comment);
+
+		/* Skip blank lines */
+		range_trim(&line);
+		if (line.len == 0)
+			continue;
 
-	email_start = memchr(line, '<', length);
-	if (!email_start)
-		return -1;
-	email_end = memchr(email_start, '>', length - (email_start - line));
-	if (!email_end)
-		return -1;
-	assert(email_end > email_start);
+		/* Get the first name and email */
+		if (!range_split(line, '<', &name_a, &line))
+			return -1; /* garbage in line */
+		if (!range_split(line, '>', &email_a, &line))
+			return -1; /* unfinished <> pair */
+
+		/* Get an optional second name and/or email */
+		two_emails = range_split(line, '<', &name_b, &line);
+		if (two_emails && !range_split(line, '>', &email_b, &line))
+			return -1; /* unfinished <> pair */
+
+		if (line.len > 0)
+			return -1; /* junk at end of line */
+
+		/* Trim whitespace from around names */
+		range_trim(&name_a);
+		range_trim(&name_b);
+
+		*found = true;
+		if (name_a.len > 0)
+			*real_name = name_a;
+
+		if (two_emails) {
+			*real_email = email_a;
+			*replace_email = email_b;
+
+			if (name_b.len > 0)
+				*replace_name = name_b;
+		} else {
+			*replace_email = email_a;
+		}
+		break;
+	}
 
-	*email_len = email_end - email_start - 1;
-	*email = email_start + 1;
-	if (*email == email_end && !allow_empty_email)
-		return -1;
+	return 0;
+}
 
-	// Now look for the name.
-	name_start = line;
-	while (name_start < email_start && isspace(*name_start))
-		++name_start;
+int git_mailmap_parse(
+	git_mailmap **mailmap,
+	const char *data,
+	size_t size)
+{
+	char_range file = { data, size };
+	git_mailmap_entry* entry = NULL;
+	int error = 0;
 
-	*name = name_start;
+	*mailmap = git__calloc(1, sizeof(git_mailmap));
+	if (!*mailmap)
+		return -1;
 
-	name_end = email_start;
-	while (name_end > name_start && isspace(*(name_end - 1)))
-		name_end--;
+	/* XXX: Is it worth it to precompute the size? */
+	error = git_vector_init(&(*mailmap)->entries, 0, NULL);
+	if (error < 0)
+		goto cleanup;
+
+	while (file.len > 0) {
+		bool found = false;
+		char_range real_name, real_email, replace_name, replace_email;
+		size_t size = 0;
+		char *buf = NULL;
+
+		error = git_mailmap_parse_single(
+			&file, &found,
+			&real_name, &real_email,
+			&replace_name, &replace_email);
+		if (error < 0)
+			goto cleanup;
+		if (!found)
+			break;
+
+		/* Compute how much space we'll need to store our entry */
+		size = sizeof(git_mailmap_entry);
+		range_copyz(NULL, &size, real_name);
```

Comment:
I'm not sure it's worth making `range_copyz` accept NULL vs just doing something like:

```c
size = sizeof(git_mailmap_entry) +
    real_name.len + 1 +
    real_email.len + 1 +
    // ...
    ;
```

Also wow, you decided to stash everything into the same allocation, your C skills are way over mine :P